File: apps/web/src/api/pmStream.ts
/* Sending a PM message and watching the turn happen.
 *
 * Lives outside `client.ts` for two reasons: that file is held at its recorded size by the
 * god-file ratchet, and this is a different shape of call — everything there resolves once with
 * a parsed body, where this one reports as it goes.
 *
 * The runs stream uses `EventSource`, which is GET-only; a chat turn is a POST with a body, so it
 * reads its own streamed response instead. Auth needs nothing special — a normal same-origin POST
 * carries the session cookie, where EventSource forced the `?token=` arrangement on the run path.
 */

import { api } from "./client";
import { apiFetch } from "./auth";
import { createSseParser } from "@/lib/sse";

/** One lookup Quincy made, live or as stored with the finished turn. */
export type PmStep = { kind: string; tool: string; arg: string; duration_ms?: number };

/** What a watcher is told while the turn runs. `text` is prose he produced mid-turn; the FINAL
 *  reply never arrives this way, because the transcript renders it. */
export type PmStreamEvent =
  | { event: "step"; data: { id?: string; kind: string; detail: string } }
  | { event: "text"; data: { text: string } }
  | { event: "done"; data: PmTurnResult }
  | { event: "error"; data: { detail: string } };

export type PmTurnResult = Awaited<ReturnType<typeof api.sendMessage>> & { steps?: PmStep[] };

/** Send a message and report the turn as it happens; resolves with the same result the plain
 *  endpoint returns.
 *
 *  Falls back to that plain endpoint whenever streaming is not actually available — no body
 *  reader, or a response that is not an event stream. The fallback is why the existing chat tests
 *  keep passing unchanged, and they are therefore its regression guard.
 *
 *  A watcher that throws is not allowed to lose the turn: the reply matters more than the
 *  animation, which is the same rule the server applies to its own step listener. */
export async function sendMessageStreaming(
  projectId: string,
  text: string,
  attachmentIds: string[],
  sessionId: string | undefined,
  onEvent: (event: PmStreamEvent) => void,
): Promise<PmTurnResult> {
  let response: Response;
  try {
    response = await apiFetch(`/api/projects/${projectId}/messages/stream`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        text,
        attachments: attachmentIds.map((attachment_id) => ({ attachment_id })),
        session_id: sessionId ?? null,
      }),
    });
  } catch {
    // The request never landed — an older server without this route, a proxy in the way. Nothing
    // ran, so sending again on the plain endpoint is safe and is the whole point of the fallback.
    //
    // Note where this catch ENDS: once bytes have been read the turn is underway on the server,
    // and retrying then would risk a second turn from one message. A failure after that point is
    // reported, never retried.
    return api.sendMessage(projectId, text, attachmentIds, sessionId);
  }

  const streaming = response.body && response.headers.get("content-type")?.includes("event-stream");
  if (!streaming) {
    return api.sendMessage(projectId, text, attachmentIds, sessionId);
  }

  const reader = response.body!.getReader();
  const decoder = new TextDecoder();
  const parse = createSseParser();
  let result: PmTurnResult | null = null;

  for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    for (const frame of parse(decoder.decode(value, { stream: true }))) {
      const typed = frame as PmStreamEvent;
      if (typed.event === "done") result = typed.data;
      if (typed.event === "error") throw new Error(typed.data.detail);
      try {
        onEvent(typed);
      } catch {
        // Watching is decoration. Losing it costs the status line; letting it throw here would
        // cost the answer.
      }
    }
  }

  if (!result) {
    // The stream ended without saying how the turn went. The work itself runs on the server and
    // finished or will; the transcript is the authority, so ask it rather than inventing a reply.
    throw new Error("the connection ended before the turn finished");
  }
  return result;
}

File: apps/api/mosaera_api/pm_stream.py
"""Watch a PM turn happen, over one streamed POST.

Deliberately not the runs' machinery. That stream is an `EventSource` over a long-lived
supervised run, with a session registry, a replay backlog and reconnect semantics — and it is
GET-only, which a chat turn (a POST with a body) cannot use at all. A conversation turn lives
about ten seconds and belongs to one asker, so it streams its own response body and keeps none of
that apparatus.

**The stream is a VIEW of the work, never the work itself.** The turn runs on a thread and writes
its rows whether or not anyone is still listening, so closing the tab loses the live display and
nothing else — the transcript already has the answer on the next fetch. Anything that made the
turn depend on the reader would trade a durable record for a nicer animation.

Async generator, not sync: the repo learned on the runs endpoint that a sync generator streamed
through `iterate_in_threadpool` pins one anyio worker per open connection, and there are about
forty. Idling here holds nothing.
"""

from __future__ import annotations

import asyncio
import json
import logging
import queue
import threading
from collections.abc import AsyncGenerator
from typing import Any

from mosaera_memory import MemoryStore

from mosaera_api.pm_turn import pm_chat

_log = logging.getLogger(__name__)

#: How often the drain wakes while the model is working. Short because a person is watching a
#: conversation, not supervising a run — the runs' 250ms would be visible as lag on a status line
#: that is meant to feel immediate. The turn is seconds long, so the poll costs little.
_POLL_SECONDS = 0.05

#: Internal marker that the worker finished. Never written to the wire.
_END = "_end"


def _frame(event: str, data: dict[str, Any]) -> str:
    """One SSE frame. Same wire shape as the runs stream, so anything that already reads SSE
    here reads this too — the difference between the two is the transport around it, not the
    bytes."""
    return f"event: {event}\ndata: {json.dumps(data)}\n\n"


async def stream_turn(
    memory: MemoryStore,
    project_id: str,
    text: str,
    attachment_ids: list[str] | None = None,
    session_id: str | None = None,
) -> AsyncGenerator[str, None]:
    """Run one turn on a thread and report it as it happens.

    Four events. `step` is a lookup starting, `text` is Quincy saying something mid-turn, `done`
    carries exactly the payload the plain endpoint returns, and `error` covers the case the turn
    itself could not handle.

    Note there is no separate `failed` event: `done`'s payload already carries `failure_cause`, so
    the client reads the outcome in one place rather than two that could disagree.

    Typed as an ASYNC GENERATOR, not an `AsyncIterator`. This `async def` yields, so a generator is
    what it is — and the difference is not cosmetic: a caller that abandons the stream mid-turn has
    to `aclose()` it to run the cleanup, and `AsyncIterator` does not carry that method. The
    abandonment test in `test_pm_stream.py` calls exactly that, which is how the mis-annotation
    surfaced.
    """
    events: queue.Queue[tuple[str, dict[str, Any]]] = queue.Queue()
    outcome: dict[str, Any] = {}

    def work() -> None:
        try:
            outcome["payload"] = pm_chat(
                memory,
                project_id,
                text,
                attachment_ids=attachment_ids,
                session_id=session_id,
                on_event=lambda kind, payload: events.put((kind, payload)),
            )
        except Exception as exc:
            _log.warning("streamed pm turn failed for %s: %s", project_id, exc, exc_info=True)
            outcome["error"] = str(exc)
        finally:
            # Always, on every path: a reader waiting on a queue nobody will fill again is the
            # one failure mode worse than an error frame.
            events.put((_END, {}))

    worker = threading.Thread(target=work, daemon=True, name=f"pm-turn-{project_id}")
    worker.start()

    while True:
        try:
            kind, payload = events.get_nowait()
        except queue.Empty:
            await asyncio.sleep(_POLL_SECONDS)
            continue
        if kind == _END:
            break
        yield _frame(kind, payload)

    if "payload" in outcome:
        yield _frame("done", outcome["payload"])
    else:
        # `pm_chat` handles its own failures and returns a payload for them, so reaching here
        # means something outside that contract broke. Say so rather than closing silently.
        yield _frame("error", {"detail": outcome.get("error", "the turn ended unexpectedly")})
